src/diagnostic/thread_local.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2024 FastLabs Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+
+use crate::diagnostic::Visitor;
+use crate::Diagnostic;
+
+thread_local! {
+    static CONTEXT: RefCell<BTreeMap<String, String>> = const { RefCell::new(BTreeMap::new()) };
+}
+
+#[derive(Default, Debug, Clone, Copy)]
+#[non_exhaustive]
+pub struct ThreadLocalDiagnostic {}
+
+impl ThreadLocalDiagnostic {
+    pub fn name(&self) -> &'static str {
+        "thread-local"
+    }
+
+    pub fn insert<K, V>(&self, key: K, value: V)
```

Comment:
Maybe remove &self?

Reply:
As far as rustc can eliminate the overhead, I'd prefer to keep the helper struct to align concepts - https://x.com/tison1096/status/1877365316621115855

Reply:
They are the same so it's easy to get a copy by `let diag = ThreadLocalDiagnostic::default()`.

Reply:
But perhaps `ThreadLocalDiagnostic::insert(k, v)` would be OK.